### DISCUSSION TITLE
Score comparison clarification

### DISCUSSION THREAD

**Discussion by mikepictor** [TimefoldAI/timefold-solver#679] [General]
I was reading https://timefold.ai/docs/timefold-solver/latest/constraints-and-score/overview, and I think I realized something I didn't realize before, but I want to confirm it.

Is a score of -1hard/-10soft "worse" than a score of -1hard/-5soft?

IE if they have the same hard penalty, does soft matter at all, will it consider the latter a better (though still bad) result, or does the -1 hard mean that the rest is ignored?

I have a scenario where I get -1medium/-30soft. In this scenario, the medium is unavoidable, the incoming data is such that it can't do better, but the soft absolutely CAN be improved, and doing so would make a bad situation "less bad". It would be making the best of a bad circumstance. Is this a hopeless case?

**Comment by triceo** [comment]
All score levels are considered. If level 0 is the same, then level 1 is used, then level 2, ... It is a total ordering, where two scores are only equal when all of their component levels are equal.

If it weren't like that and only the most significant level mattered, local search would have had a much harder time finding a good solution. This is an important thing to understand - if two scores are the same, the solutions they represent will be considered the same; this situation is often called a score trap.

For this reason, you should write as granular constraints as you possibly can. (Example: instead of penalizing a task which is overdue by a flat `1`, penalize by the amount of minutes that the task is overdue.)

**Reply by mikepictor** [reply]
Yes, I do that, but in this example I wanted to be sure that -1medium/0soft is "better" than -1medium/-30soft

It sounds like you are saying it is. I read the text under the link I quoted differently

In that case...I can't figure out why my soft score isn't better, the planning entity should have had all it needs to improve the soft score, but I'll have to track that down I guess. My worry was that the -1medium was causing it to not even attempt to improve the soft score.
